### PR TITLE
feat: Make headers to add to request in openid-connect plugin configurable.

### PR DESCRIPTION
**Pull request by jenskeiner**
### What this PR does / why we need it:
See #2880. The current implementation may lead to very large headers, thus bloating the request size to downstream substantially. This can be an issue by itself but is also not desired if the different tokens (access, ID, user info) largely contain the same information. Also, instead of using the `X-Access-Token` header it may be more appropriate to use the `Authorization` header for the access token, so that downstream plugins or services can extract it correctly; see e.g. `authz-keycloak` plugin. Finally, the plugin was so far only expecting an incoming access token in the `Authorization` header, but not the `X-Access-Token` header, which seemed somewhat inconsistent.

I've added options that can be configured for the `openid-connect` plugin that allow to control which headers get added and which header specifically shall be used for the access token. The defaults reproduce the current behavior. The plugin will also look in the `X-Access-Token` header, in addition to the `Authorization` header, for an access token in an incoming request.

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
Haven't added any new cases for now. The headers set by the plugin were not tested, as far as I can tell. It may be a good idea to think about some new tests around here. That would likely require a bit of legwork to set up a dummy OIDC Identity Provider with all needed endpoints to go through all possible code paths though.
* [x] Have you modified the corresponding document?
I understand that the plugin documentation should be updated, but I want to get some feedback on the code changes first. Happy to adjust the documentation if the code changes get signed off.
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:169]: Why you call the boolean value `has_token`?

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:155]: Why remove this part?

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:196]: `add_access_token_header(ctx, conf, token)`
Argument mismatch.

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:246]: Should check if `response.access_token` is nil

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:17]: `local string = string`
is enough.

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:45]: Let's use the new way to set the default value:
https://github.com/apache/apisix/blob/9dfe697b644691509b55cd7ec58188b33dd3e792/apisix/plugins/cors.lua#L37

**Review by spacewander — CHANGES_REQUESTED** [review]

**Comment by spacewander** [comment]
Please write new tests to cover your change, and make sure they are passed.

File: apisix/plugins/openid-connect.lua
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua]: It's a bug, have pushed a fix.

File: apisix/plugins/openid-connect.lua
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:17]: ok

File: apisix/plugins/openid-connect.lua
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:45]: Sure, hadn't seen it before.

Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:45]: Sure, hadn't seen it before. Will change.

File: apisix/plugins/openid-connect.lua
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:196]: Correct, good spot. Will fix.

File: apisix/plugins/openid-connect.lua
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:155]: The reason is that adding the user info header is now handled in the `introspect` method. The header will still be added, if so configured, and introspection is done via the actual endpoint as opposed to just checking against the public key. Let me know if you see any problem with this approach.

File: apisix/plugins/openid-connect.lua
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:246]: Agreed. Will change.

File: apisix/plugins/openid-connect.lua
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua]: I've changed the scheme for the options that I've added. But have left the existing options untouched to not make this PR any larger.

**Comment by jenskeiner** [comment]
> Please write new tests to cover your change, and make sure they are passed.

I'm on it. I had a look at the existing structure and I'm relatively confident I can add more test cases to verify the setting of the respective headers.

However, one thing is still unclear to me: There are cases already where token introspection is performed against a configured identity provider endpoint. The cases were added by @sshniro here: https://github.com/apache/apisix/commit/14f979edf84c5a76a96902cc2ab8e5697a905fe9

Can anyone point me to the place where this introspection endpoint is set up for the tests? I can't find the corresponding code inside the project.

I may need to add more dummy endpoints to verify the authorization flow in case a bearer token is not present in the incoming request and I want to make that consistent with the existing dummy introspection endpoint.

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:203]: Token is invalid if has_token is false but conf.bearer_only is true

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:220]: Ditto

**Review by spacewander — CHANGES_REQUESTED** [review]

**Comment by sshniro** [comment]
Hi @jenskeiner for your question on the introspection endpoint test case, I have used the Keycloak to test this scenario.

The code link to setup Keycloak:
https://github.com/apache/apisix/blob/f84a44e409826cfbeebf89ec7a8ebe47dcf9e387/.travis/linux_openresty_runner.sh#L27
Test cases written for the introspection with Keyclaok:
https://github.com/apache/apisix/blob/f84a44e409826cfbeebf89ec7a8ebe47dcf9e387/t/plugin/openid-connect.t#L424

@membphis the docker image is using my personal docker repo, if needed I can migrate it to an official APISIX docker repository.

**Comment by membphis** [comment]
@membphis the docker image is using my personal docker repo, if needed I can migrate it to an official APISIX docker repository.

seems good for me.

what do you think? @moonming

**Comment by moonming** [comment]
> @membphis the docker image is using my personal docker repo, if needed I can migrate it to an official APISIX docker repository.
> 
> seems good for me.
> 
> what do you think? @moonming

we can add `keycloak-apisix` docker to CI, and I am not sure I am not sure if it is appropriate to migrate it to an official APISIX docker repository, most of APISIX users don't need keycloak.

File: t/plugin/openid-connect.t
Comment (membphis) [inline t/plugin/openid-connect.t:115]: `8888`, Is it a valid "openid" port?

Comment (jenskeiner) [reply t/plugin/openid-connect.t:115]: Hey, I'm trying to add tests to check if the plugin sets the headers correctly as configured. I don't have a local development setup yet, so have to run the checks on this PR on GitHub to see if it's working.

I switched the upstream port to `8888` since that's where a `mendhak/http-https-echo` container is running. This one returns in the response body the full content of the incoming request as a JSON string. So I think I can check if the plugin has set the headers correctly by checking the response body. If the tests still work with the new port, I will try to add a case to do some actual verification of the headers.

Comment (spacewander) [reply t/plugin/openid-connect.t:115]: @jenskeiner 
You can setup the test environment via https://github.com/apache/apisix/blob/master/.travis/linux_openresty_runner.sh

Comment (jenskeiner) [reply t/plugin/openid-connect.t:115]: Yes, thanks for the link. It's more that I have to set up a fresh Linux VM or similar since I want that separate from the rest.

Comment (spacewander) [reply t/plugin/openid-connect.t:115]: And you can write your echo-back here: https://github.com/apache/apisix/blob/master/t/lib/server.lua

Comment (jenskeiner) [reply t/plugin/openid-connect.t:115]: Understood, seems to be the better option. Will adjust the code.

File: t/plugin/openid-connect.t
Comment (tokers) [inline t/plugin/openid-connect.t:631]: Why using `9z`?

Comment (jenskeiner) [reply t/plugin/openid-connect.t:631]: I'm in the middle of the process of adding cases, so will fix the numbering when done. No point doing it now as I don't know how many cases I will add.

Also, why number the cases in the first place? If you insert just one somewhere you have to adjust the numbers of all cases below.

Comment (jenskeiner) [reply t/plugin/openid-connect.t:631]: Also, if anyone could help me setting up a proper dev environment so I don't have to go through the entire suite of GitHub checks all the time just to run the tests on the oidc module, I would appreciate that very much.

I tried setting up an Ubuntu 18.04 VM according to Travis configuration, but it's frankly a nightmare to get everything up and running because most scripts need tiny little adjustments to actually work as expected in that environment. In the end, I seemingly managed to get all containers and everything up and running, but the tests still wouldn't run because the APISIX process was still throwing up a lot in the error log. Not a single test passed as a result.

I need to find a better development workflow or I may have to abandon this whole ticket. At this stage, I'm wasting too much time on stuff not really related to the actual dev work. Can't afford that any longer.

Maybe it might be a good idea to set up a Vagrant box or similar so new developers don't have to go through all this trial and error.

Any hints appreciated.

Comment (spacewander) [reply t/plugin/openid-connect.t:586]: It is amazing that you finally get all the test passed! Thanks a lot!

Comment (spacewander) [reply t/plugin/openid-connect.t:586]: @jenskeiner 
We can fulfill the tests for you later. Before we can merge, the only thing you need to do is to update the English doc: https://github.com/apache/apisix/blob/master/doc/plugins/openid-connect.md#attributes

Comment (jenskeiner) [reply t/plugin/openid-connect.t:586]: Hang on, I disabled most of the tests, so currently just running the OIDC ones. Need to revert that before we proceed.

Also, was thinking how to test the path where the initial request doesn't have a token and is redirected to the identity provider. When the user has authenticated, the identity provider will call the redirect URL with a code so that APISIX (the relying party) can obtain access token, ID token, and userinfo from the provider's endpoints.

We need to test this because the plugin wil set headers to the upstream service accordingly in this case.

Probably best to mock up the endpoints instead of interacting with the Keycloak container. I'll see if I can get this done as well.

In any case, I will post here when I believe the PR is actually ready for final review round.

**Comment by jenskeiner** [comment]
@spacewander @membphis @moonming Sorry for the noise, but took a while to set up test cases that go through the full OIDC Relying Party flow which can be used when the incoming request doesn't contain a bearer token yet.

Please give this another good review and let me know any issues. Thanks.

I know, I still need to update the documentation, but will wait on feedback regarding code changes first.

**Comment by jenskeiner** [comment]
The code changes in this PR are basically just to solve some issues I was seeing when combining this plugin with the `auth-keycloak` one. I wanted to make sure that the latter can extract the access token from the request, even if it is contained in a session cookie or potentially some other server-side storage as supported by the `openidc` module. The changes make sure that the token can always be added in a header where `authz-keycloak` expects it. I also wanted to reduce some noise from the ID token and the user info that were getting added in separate headers because you probably don't need those always and they make the HTTP header even larger.

Generally, I think the interplay between this plugin and `auth-keycloak` should be re-considered. Messing with the request headers could be avoided completely if `auth-keycloak` could leverage the `openidc` module to extract the access token from the request (from a header or cookie or...) itself using a method that is currently internal to that module. Then, it wouldn't need to rely on the token being stored in a header and code to extract the token could be removed in favor of using the `openidc` module's code for that.

I'm just wondering what anyone's thoughts on this are.

The current solution with the headers works for me and I'll look at `authz-keycloak` next to start the work on lazy path loading from Keycloak. But longer term, would be better to clean up the code some more and not rely on the `openid-connect` plugin to fix up the headers for the `auth-keycloak` plugin. Thoughts and feedback welcome.

@sshniro Maybe you have thoughts on this as well...

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [inline apisix/plugins/openid-connect.lua:156]: Better to check the err: `_`.

**Review by spacewander — CHANGES_REQUESTED** [review]

File: t/plugin/openid-connect.t
Comment (jenskeiner) [reply t/plugin/openid-connect.t]: Fixed. Thanks.

File: apisix/plugins/openid-connect.lua
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua]: Done. Added a test case for a malformed Authorization header as well.

File: apisix/plugins/openid-connect.lua
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:218]: @spacewander Can you specify what you suggest should be checked? The `add_access_oken_header` method is only called when an access token is actually available in the request.

**Review by sshniro — APPROVED** [review]

File: apisix/plugins/openid-connect.lua
Comment (spacewander) [reply apisix/plugins/openid-connect.lua:218]: @jenskeiner 
There is two checks outside this function. I think we can move the check inside so that we don't need to repeat it.

Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:218]: Understood and done.

**Review by spacewander — APPROVED** [review]

**Comment by spacewander** [comment]
Fantasy! Before we can merge it, @jenskeiner please write some English docs.

**Comment by jenskeiner** [comment]
> Fantasy! Before we can merge it, @jenskeiner please write some English docs.

Yep, will do, may take until next Monday though. In the meantime, let me know any worries.

**Comment by jenskeiner** [comment]
@spacewander Updated the documentation now as well. Can you give it another good review to ensure it's consistent and clear enough?

**Comment by spacewander** [comment]
@jenskeiner 
Merged. Thanks!

**Pull request merged**
